File: 백준/Silver/2961. 도영이가 만든 맛있는 음식/도영이가 만든 맛있는 음식.py
from sys import stdin, maxsize
input = stdin.readline

n = int(input())
material = [tuple(map(int,input().split())) for _ in range(n)]

def bt(s,b,cur):
    global ans
    ans = min(ans,abs(s-b))
    for i in range(cur+1,n):
        if not vis[i]:
            vis[i] = True
            bt(s*material[i][0],b+material[i][1],i)
            vis[i] = False
    
ans = maxsize
vis = [False] * n
for i in range(n):
    vis[i] = True
    bt(material[i][0],material[i][1],i)
    vis[i] = False

print(ans)

File: 백준/Silver/2961. 도영이가 만든 맛있는 음식/도영이가 만든 맛있는 음식.java
import java.io.*;
import java.util.*;

class Main {

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		int n = Integer.parseInt(br.readLine());
		int ans = Integer.MAX_VALUE;
		int[] v1 = new int[n];
		int[] v2 = new int[n];
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			v1[i] = a;
			v2[i] = b;
		}
		for (int j = 1; j < (1 << n); j++) {
			int t1 = 1;
			int t2 = 0;
			for (int k = 0; k < n; k++) {
				if (((1 << k) & j) > 0) {
					t1 *= v1[k];
					t2 += v2[k];
				}
			}
			ans = Math.min(Math.abs(t1 - t2), ans);
		}
		System.out.println(ans);
	}

}